src/azure-cli-core/azure/cli/core/_profile.py

```diff
@@ -1008,6 +1029,29 @@ def retrieve_token_for_user(self, username, tenant, resource):
             self.persist_cached_creds()
         return (token_entry[_TOKEN_ENTRY_TOKEN_TYPE], token_entry[_ACCESS_TOKEN], token_entry)
 
+    def retrieve_ssh_certificate_for_user(self, tenant, modulus, exponent, refresh_token):
+        """
+        This is added only for vmssh feature.
+        It is a temporary solution and will deprecate after adoption to MSAL completely.
+        """
+        from azure.cli.core._msal import SSHCertificateClientApplication
+        scopes = ["https://pas.windows.net/CheckMyAccess/Linux/user_impersonation"]
```

Comment:
will this be different under different cloud env?

Reply:
Currently there is only one link. This was described in [this document](https://microsoft.sharepoint.com/:w:/r/teams/AutomationConfigAll/_layouts/15/guestaccess.aspx?rtime=MB8KjvQD10g&share=EXj-_T0zafdLpB63hyi82fUBLnUfp9t-isLgaNOh6KTzCQ). The document says it may changed to `sshservice.azure.net` in the future.

Reply:
This feature currently seems only supported in public cloud [https://docs.microsoft.com/en-us/azure/active-directory/devices/howto-vm-sign-in-azure-ad-windows#supported-azure-regions-and-windows-distributions](https://docs.microsoft.com/en-us/azure/active-directory/devices/howto-vm-sign-in-azure-ad-windows#supported-azure-regions-and-windows-distributions), but we should add this endpoint in the cloud.py and throw error for non-public cloud

---
In reply to: [412720221](https://github.com/Azure/azure-cli/pull/12999#discussion_r412720221) [](ancestors = 412720221)

Reply:
I think maybe you should let the client pass scopes and data to get token to make this method more generic, this method contain too much implementation detail for SSHcert

---
In reply to: [413537394](https://github.com/Azure/azure-cli/pull/12999#discussion_r413537394) [](ancestors = 413537394,412720221)

Reply:
> I think you should let the client pass scopes and data to get token to make this method more generic, this method contain too much implementation detail for SSHcert
> 
> In reply to: [413537394](https://github.com/Azure/azure-cli/pull/12999#discussion_r413537394) [](ancestors = 413537394,412720221)

Agreed. I will change it.